mypy/checker.py

```diff
@@ -1618,12 +1628,12 @@ def check_compatibility(self, name: str, base1: TypeInfo,
             return
         first = base1[name]
         second = base2[name]
-        first_type = first.type
-        if first_type is None and isinstance(first.node, FuncBase):
-            first_type = self.function_type(first.node)
-        second_type = second.type
-        if second_type is None and isinstance(second.node, FuncBase):
-            second_type = self.function_type(second.node)
+        if isinstance(first.node, TypeInfo) and isinstance(second.node, TypeInfo):
+            # allow nested classes with the same name
```

Comment:
In general I agree it is better to admit this little unsafety and fix a common pain point.

Please add a more detailed comment with motivation/trade-offs. Here are some points to mention:
* This is technically unsafe due to possible errors when accessing members of the nested classes.
* But this creates many false positives for Django.
* Also mypy anyway doesn't flag this for single inheritance.
* A potential solution may be to require at least structural compatibility (since requiring nominal one is impractical).